osrf/lrauv
# Break up #70 - visualization only, no lat/long to Cartesian conversion

Breaking up part of #70 into a smaller PR. This PR has all the visualizations, no lat/long to Cartesian conversion.

Looking at the files in #70, the only file that needs breaking up is `lrauv_ignition_plugins/src/ScienceSensorsSystem.cc`.
Most other files are GUI and `VisualizePointCloud.cc`, which @chapulina and I already looked at closely in #85, which has been merged into #70.

Diff of this PR from #70:
- In `buoyant_tethys.sdf`, the data file is changed to `simple_test.csv`, since lat/long to Cartesian conversion isn’t taking place, we cannot test the actual data files.
- A few parameters (`MINIATURE_SCALE`, `dimFactor`, `RES_*`) are tweaked to this test file, so that >0 markers show up and are big enough to see, with the unconverted lat/long. Bug fix on calculating `renderEvery`.

(I did a subtractive instead of additive diff :sweat_smile: I mean, it is Thanksgiving eve, and the only lines that needed deleting have to do with the lat/long to Cartesian conversion, which is <100 lines. Since we squash-merge, the diff shouldn't look weird in the commit history.)

## To test

```
ign gazebo -v 4 buoyant_tethys.sdf
```
Unpause, enable Visualize Point Cloud plugin, click refresh button. Change the data type.
Check that for chlorophyll, 13 markers are visualized, because the 17 others are NaN. For salinity and temperature, 30 are rendered.

![2021-11-25-000555_2400x2000_scrot](https://user-images.githubusercontent.com/7608908/143382847-021ea393-19bc-48c3-8476-c3cfbb942af4.png)

![2021-11-25-000558_2400x2000_scrot](https://user-images.githubusercontent.com/7608908/143382863-30648fbf-5889-4fd4-9c4d-b851fc1d492f.png)

![2021-11-25-000601_2400x2000_scrot](https://user-images.githubusercontent.com/7608908/143382864-7ff5f28f-3bb1-4bdb-954d-17122815f6ce.png)

## Bug

When switching from salinity (30 valid points) or temperature (30 valid points) to chlorophyll (13 valid points), there is a bug. More than 13 markers are rendered for chlorophyll. Debug printout says 30 rendered in the first round (but you see 25 in the scene), then the printout goes down to 13. But since no unsubscription is performed in between, you still see 25.

![2021-11-25-000603_2400x2000_scrot](https://user-images.githubusercontent.com/7608908/143382875-895ed08e-d15b-4e76-ad49-885fe3374505.png)

This could be a timing issue. When a new data type is subscribed, the new float array data may not have been received before it tries to render the markers, so it’s rendering the old data. One guess is this may have been introduced by changing to use a single array to represent all the data types. There are more edge cases that need to be taken care of when switching the data in that array (I saw some of these unclean cases when I was on a time crunch, which was the reason I went to multiple arrays).

We may not have noticed this before because the real data has too many points to really test this. The simple test shows this very clearly.

I tried tentatively moving the mutex in `PublishMarkers()`, to see if it’s a lucky hit since `nPts` takes the size of the point cloud. But that didn’t fix it.

@chapulina could you help looking into this behavior?
